Improve support for Swift-only initializers in @_objcImplementation checker

beccadax opened the pull request:
The typical rules for member implementations don't work very well for initializers:

* Initializers cannot be made `final`, so we need to support `@nonobjc` as an opt-out keyword for them.
* Designated and required initializers don't function correctly without dynamic dispatch, dynamic dispatch of Swift-only members requires a vtable, and `@_objcImplementation` classes don't have a vtable, so it shouldn't be possible to define Swift-only designated and required initializers on an `@_objcImplementation` class. (Swift-only `convenience` inits and `@objc` inits of any kind are okay.)

Enforce these new rules, and additionally make several related improvements to `@_objcImplementation` diagnostics:

* Make the fix-it suggesting `private` on an `@objc` member replace an existing access control keyword if one is present.
* Add a separate `ObjCReason` for members that are implicitly `@objc` because they belong to an `@_objcImplementation` extension; this improves the wording of some diagnostics and allows us to avoid emitting redundant diagnostics about members that aren't valid as `@objc`.
* Add a fallback diagnostic about `@_objcImplementation` extension members that require a vtable entry. This would have caught the designated/required init issue even without knowing the specific cause of the problem.

Fixes rdar://109121293.

This PR will probably be easier to review commit-by-commit.

beccadax committed:
Replace existing access control in objcImpl fixit

The fix-it suggesting that an unmatched `@objc`-able member can be turned `private` always suggested adding a new modifier, even if one already existed. Make it suggest replacing the existing one instead.

beccadax committed:
Allow @nonobjc inits in objcImpl extensions

Initializers can’t be made final, but they *can* be made @nonobjc. (This isn’t always enough to ensure they’re actually valid, but I’ll get to that in a follow-up commit.)

beccadax committed:
Add ObjCReason for @objcImpl

Besides improving diagnostics, this also allows us to track whether there were any @objc failures using the invalid bit on the @_objcImplementation attribute.

beccadax committed:
Diagnose invalid Swift-only @objcImpl inits

Required and designated inits have to be overridable—the former so that uses of the initializer on `any T.Type` will call the subclass initializer, the latter so that inherited convenience inits will call the subclass initializer. However, Swift-only members use vtables to dispatch to subclasses, and @objcImpl classes don’t have vtables, so their Swift-only inits cannot be made overridable. Upshot: Swift-only inits on @objcImpl classes must be `convenience`, not `required` or designated.

Enforce this rule in the ObjCImplementationChecker.

Fixes rdar://109121293.

beccadax commented:
@swift-ci please test

tshortli commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1841,6 +1841,25 @@ NOTE(objc_implementation_requirement_here,none,
+
+// Fallback diagnostic; super-general by nature.
+ERROR(objc_implementation_member_requires_vtable, none,
+      "%kind0 is not valid in an '@_objcImplementation' extension because it "
```
I take it from the comment that we want to avoid emitting this if possible, and in the test case that inspired this we emit a better, more actionable diagnostic (which is great!) I noticed that after that change, though, we no longer seem to have any test case that causes this diagnostic to be emitted. Is it possible to add one? I'm also wondering if there is a way we can word this diagnostic that's friendlier to folks who don't understand what a vtable is, or even if they do understand what one is, wouldn't know what to do to address the problem. I realize that may be heavily dependent on the member in question, though.

beccadax replied:
If I knew of another way to cause this diagnostic, I would add another special case. Short of a debug flag that disables checking for all of the special cases, I don’t see a good way to test this.

I’m willing to consider wording the diagnostic differently. Maybe something like:

```
      "%kind0 is not valid in an '@_objcImplementation' extension because it "
      "is an overridable Swift-only %kindonly0",
```

beccadax replied:
I've updated the wording, and also rebased on post-rebranch `main`.

beccadax committed:
Add @objcImpl vtable fallback diagnostic

Nothing in an `@_objcImplementation` block—including Swift-only declarations—should ever require a vtable entry. Diagnose any such members.

The diagnostic emitted here is intended to be a fallback that will be used when a vtable entry is needed but we don’t know the specific reason. A future commit will add a more specific diagnostic for Swift-only non-convenience inits.

beccadax commented:
@swift-ci please smoke test and merge
